go-vet: add support for -shadow and -shadowstrict options

kostya-sh opened the pull request:
Starting from Go 1.6 `go vet -all -shadow' means all default checkers and also
shadow checker. This allows to add options to enable shadow check.

The discussion in #765 has more details.

lunaryorn commented:
Thank you.  Please add documentation for these options in `doc/languages.texi`.

lunaryorn commented on flycheck.el:
```diff
@@ -6562,14 +6562,36 @@ take an io.Writer as their first argument, like Fprintf,
             (option "-printfuncs=" flycheck-go-vet-print-functions concat
                     flycheck-option-comma-separated-list)
+            (option-flag "-shadow" flycheck-go-vet-shadow)
+            (option-flag "-shadowstrict" flycheck-go-vet-shadowstrict)
```
What is the difference between `-shadow` and `-shadowstrict`?

lunaryorn commented on flycheck.el:
```diff
@@ -6562,14 +6562,36 @@ take an io.Writer as their first argument, like Fprintf,
+
+When non-nil, enable `go tool vet' shadowed variables check.
+
+This option requires Go 1.6 and older."
```
Older?  Shouldn't this be "Go 1.6 or *newer*"?

lunaryorn commented on flycheck.el:
```diff
@@ -6562,14 +6562,36 @@ take an io.Writer as their first argument, like Fprintf,
+When non-nil, make `go tool vet' shadowed variables check more
+strict; can be noisy.
+
+This option requires Go 1.6 and older."
```
Likewise.

lunaryorn commented:
By the way, can't we merge this option into a single one that takes three values, `nil`, `t` (`-shadow`) and `strict` (`-shadow -shadowstrict`)?

kostya-sh commented:
Single option is good idea. I've implemented it, addressed your other comments and updated the PR. Please let me know if anything else could be improved or should be changed.

lunaryorn commented on flycheck.el:
```diff
@@ -6562,14 +6562,29 @@ take an io.Writer as their first argument, like Fprintf,
             (option "-printfuncs=" flycheck-go-vet-print-functions concat
                     flycheck-option-comma-separated-list)
+            (option-flag "-shadow" flycheck-go-vet-shadow)
+            (eval (when (eq flycheck-go-vet-shadow 'strict) "-shadowstrict"))
```
Nice.

lunaryorn commented on flycheck.el:
```diff
@@ -6562,14 +6562,29 @@ take an io.Writer as their first argument, like Fprintf,
 See URL `http://golang.org/cmd/go/' and URL
 `http://godoc.org/golang.org/x/tools/cmd/vet'."
   :command ("go" "tool" "vet"
+            "-all"
```
Why's that?  I don't think that was in the first version of your PR?

kostya-sh replied:
You will probably complain about weirdness of go command line tools again, but this is necessary ;)

`vet -shadow` runs only shadow check
`vet -all` runs all checks apart from shadow check (same as `vet` with no options)
`vet -all -shadow` runs all checks including shadow

`-shadowstrict` is a flag that is only used by shadow check.

lunaryorn replied:
Well, that is weird.  In other words, `go vet` and `go vet -all` are identical?

kostya-sh replied:
Yes. Internally `-all` is a flag that is set to `true` by default. Full docs: https://godoc.org/golang.org/x/tools/cmd/vet

lunaryorn commented:
@kostya-sh I cherry-picked your change onto `master`, you may safely delete your branch.  Thank you very much for your contribution, and for your patience with my questions and remarks :+1:.

kostya-sh commented:
@lunaryorn, thank you!

adamcohen commented:
Just came across this @kostya-sh and wanted to say thank you for adding this option, I've run into this issue so many times, it's great to finally have a solution!!

Also, thanks to @lunaryorn for your work on flycheck, it's helped save me hours of frustration!
